rust/serving/src/app.rs

```diff
@@ -73,34 +74,58 @@ where
     Ok(())
 }
 
+#[derive(Clone)]
+/// New type to store the TID of a request in Axum's request extensions
+struct Tid(String);
+
 pub(crate) async fn router_with_auth<T, U>(app: AppState<T, U>) -> crate::Result<Router>
 where
     T: Clone + Send + Sync + DataStore + 'static,
     U: Clone + Send + Sync + CallbackStore + 'static,
 {
     let layers = ServiceBuilder::new()
+        // Ensure all requests has the TID header set.
+        // This is used in the tracing layer to define the top-level span for a request.
+        .map_request({
+            let tid_header = HeaderName::from_bytes(app.settings.tid_header.as_bytes()).unwrap();
+            move |mut req: Request<Body>| {
+                if ["/metrics", "/readyz", "/livez", "/sidecar-livez"].contains(&req.uri().path()) {
+                    return req;
+                }
+
+                let tid = match req.headers().get(&tid_header) {
+                    Some(tid) => String::from_utf8_lossy(tid.as_bytes()).to_string(),
+                    None => Uuid::now_v7().to_string(),
+                };
+                req.extensions_mut().insert(Tid(tid));
+                req
+            }
+        })
         // Add tracing to all requests
         .layer(
             TraceLayer::new_for_http()
-                .make_span_with(move |req: &Request<Body>| {
-                    let req_path = req.uri().path();
-                    if ["/metrics", "/readyz", "/livez", "/sidecar-livez"].contains(&req_path) {
-                        // We don't need request ID for these endpoints
-                        return info_span!("request", method=?req.method(), path=req_path);
-                    }
+                .make_span_with({
+                    move |req: &Request<Body>| {
+                        let req_path = req.uri().path();
+                        if ["/metrics", "/readyz", "/livez", "/sidecar-livez"].contains(&req_path) {
```

Comment:
There is nothing wrong in having a span for health checks. Do you see any concerns?

Reply:
Currently I removed special handling of these  endpoints. So the logs for a single request will look like:
```
2025-03-19T04:12:55.727466Z  INFO request{tid="0195ac9a-b82f-7a41-88dc-6821771f674b"}: serving::app: Received request method=GET path="/health" matched_path="/health"
2025-03-19T04:12:55.727601Z  INFO request{tid="0195ac9a-b82f-7a41-88dc-6821771f674b"}: serving::app: status=200 latency=159.874µs
```

Should I make them at debug level ?

Reply:
Yeah you can move it to debug

Reply:
I tried to make it log at debug level. The first line can be made to debug level by inspecting the path, but where the second line is emitted, we can't access request path:
```rust
.on_response(
    |response: &Response<Body>, latency: Duration, _span: &Span| {
        if response.status().is_server_error() {
            // 5xx responses will be logged at 'error' level in `on_failure`
            return;
        }
        info!(status=?response.status(), ?latency)
    },
)
```
If the first line is made to debug level, the second line will only have latency and status (but doesn't contain the request path).
So I'm not making any changes related this for now.